### ISSUE TITLE
Add more configuration parameters to espeak-ng?

### ISSUE THREAD

**Issue by sthibaul** [brailcom/speechd#882]
**Is your feature request related to a problem? Please describe.**

from @albertotirla :

“the espeak-ng module doesn't even support proper inflection, same for rate boost”

“spd doesn't allow the client application to pass arbitrary parameters to the synth and the synth to report an error if those make no sense or just a warning and ignoring them”

**Describe the solution you'd like**

It's not exactly clear how we want to fix this.

- For some parameters which also make sense to other modules, we can introduce them more generally, and they can appear in the Orca configuration panel
- For other parameters which are really specific to espeak-ng, we may not want to clutter the Orca configuration panel with yet more knobs. Is it fine to just add support for them into `espeak-ng.conf`?

**Comment by sthibaul** [comment]
From @albertotirla :
“not only for espeak-ng, custom per-synth parameters should be supported for all modules, and for backwards compatibility, modules which don't define them should resort to an empty dictionary or something like that”

**Comment by albertotirla** [comment]
I believe there should be per-module configuration parameters, listed in the module's configuration file. Then, a new command should be made, which essentially indicates to the running module which custom parameters were specified and at what values, essentially a list of key->value pairs. The speech dispatcher->screenreader interface to this would be an SSIP command which returns a similarly formatted object to the one given in the module command case, the only difference being that default values provided in the configuration files are given instead of the ones set by spd when initializing the module. Then, the type of values can probably be inferred from what's given in that dictionary, since we don't expect arbitrarily complex structures as values, while we could have keys always be strings.

Alternatively, we could have a system where any key->value pair is accepted and passed on to the module, but then the module can respond with an error, say, no such parameter.

**Comment by sthibaul** [comment]
so you mean that in the orca interface we would add an area where arbitrary key/value pairs can be recorded?

**Comment by albertotirla** [comment]
@sthibaul precisely. With the possible restriction that the module specifies upfront, in its configuration file, what accepted keys are, and how the value should be like, either an actual value, or a formatted string as in sprintf stile, or a dbus signature stile one, if you think we should support a large variety of arbitrarily complex values. For our sanity, I think keys should be restricted to strings though.

**Comment by sthibaul** [comment]
it'll be much simpler to have string/string, and the module will cope with checking what is in there, indeed

**Comment by sthibaul** [comment]
that being said, in what was said above I guess some parameters would make sense to be added more globally, because they are not specific to espeak-ng ? e.g. inflection and rate boost

**Comment by albertotirla** [comment]
yep, true.
also, I just had an idea, dk how good it is
you mentioned that an aria should be introduced in orca, which would allow the user to type key value pairs. What if we could enrich that information, by hinting to orca, or odilia, or whatever other such app, what type of parameters those are, and what range they go between if applicable?
for example, if rate_boost would be a synth specific parameter and it would be of type boolean, orca would choose perhaps to implement it as a checkbox. For a range, a slider, while for a value group, a combobox. What do you think? is that too much complexity? I know though, that would make user's lives easier, because they wouldn't have to type stuff as much

**Comment by sthibaul** [comment]
That becomes more complex of course. The question is whether this complexity is worth it. As in: default parameters are supposed to be already good. If they are not, they have to be fixed. Non-default parameters are for more advanced uses, and it makes some sense to have to tinker a bit to change them. That being said, simple boolean and ranges would probably be plenty to support various kinds of needs.

**Comment by albertotirla** [comment]
so then, how should this information look in the config file? maybe dbus signature stile, or a json object with the type information and then a default value, or something simpler?

**Comment by sthibaul** [comment]
which config file are you thinking of? Orca or spd?

For spd, the module will have to know how to tune the parameter anyway, so the parameter does not need to show up explicitly in the config file (even if it'll be useful to add one so one can also configure it there).

For orca, it's for Joanie to implement as she likes it:)
